src/gui/fold_until_stop.rs

```diff
@@ -0,0 +1,77 @@
+use std::{io::stdout, sync::mpsc, thread::spawn};
+
+use crate::gui::error::GrapeGuiError;
+
+use crossterm::{
+    event::{self, KeyEventKind},
+    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
+    ExecutableCommand,
+};
+
+use ratatui::{
+    prelude::*,
+    widgets::{block::Title, *},
+    Terminal,
+};
+
+enum ThreadMessage {
+    Stop,
+}
+
+pub fn fold_until_stop<F, T>(init: T, f: F) -> Result<T, GrapeGuiError>
```

Comment:
Ah, this is "fold this function `f` over each value when it's asynchronously received until we stop"! Very nice